### ISSUE TITLE
Review

### ISSUE THREAD

**Issue by 33Domenico** [michepaolo/CI2024_lab2#1]
The code addresses the travelling salesman problem by combining a greedy approach (fast algorithm) with a genetic algorithm (slow algorithm) to find optimised paths. The use of mutation and crossover functions is well thought out, especially with the implementation of inversion_mutation, which preserves significant sections of genes by inverting their order. This approach is effective in maintaining good solutions while avoiding altering the pathways found too much.
Inversion_over_crossover is also interesting because, unlike other crossovers, it does not drastically change genes and preserves effective sections of the pathway. In addition, the choice of dynamically varying the probability of using inver_over_crossover versus mutation and vice versa is remarkable, maintaining a good balance between exploration and exploitation of the solution.
The addition of a local_search on the first 10 individuals further improves the solutions, avoiding letting the algorithm stagnate. In addition, controlling the best_cost_unchanged variable, which triggers the random reintroduction of new individuals when no improvement is observed, is a good attempt (surely there are better ones) to reshuffle the population and overcome any blockages caused by lack of diversity. However, since parental selection only takes place among the first half of the population, there is a risk that these new random solutions, due to their high initial cost, are never really considered, thus limiting the contribution to population diversity.
In summary, the combined approach of greedy, targeted crossover, local search and dynamic management of crossover and mutation probability led to excellent results, balancing solution quality and genetic diversity effectively. Excellent work!
